README.md
# aelf-bridge

English | [中文](./README.zh-CN.md)

## Table of Contents
  - [Introduction](#introduction)
  - [Installation](#installation)
    - [Using version management tools](#using-version-management-tools)
    - [Using the script tag](#using-the-script-tag)
  - [Usage](#usage)
    - [Introduction](#introduction)
    - [Initialization](#initialization)
      - [Options](#options)
    - [Get wallet account information](#get-wallet-account-information)
    - [Call contract method (read-only and send transaction)](#call-contract-method-read-only-and-send-transaction)
    - [Calling the chain API](#calling-the-chain-api)
    - [disconnect](#disconnect)

## Introduction

Since dApps are not allowed to store any wallet information, the wallet application stores AElf wallet information and can communicate directly with the AElf chain. In order to protect wallet information and provide dApps with the ability to interact with the chain, aelf-bridge can be used for interacting with the wallet.

The wallet App described here includes a mobile (iOS/Android) native app, desktop app and more.

## Installation

AElf-bridge is part of the AElf ecosystem. Since dApps are mostly web applications, we provide a `JavaScript` SDK available using `Npm` or `yarn` as a version management tool or directly with the script tag.

### Using version management tools

```bash
npm i aelf-bridge
// or
yarn add aelf-bridge
```

### Using the script tag

```html
<script src="https://unpkg.com/aelf-bridge@latest/dist/aelf-bridge.js"></script>
```

## Usage

### Introduction

The communications between a dApp and the chain need to go through some wallet software. This wallet software could be any client that has implemented the AElf bridge protocol. at the time of writing (2020.06), AElf mobile wallet App and `aelf-command` have implemented this protocol.

Since dApps are mostly web applications and web applications can communicate with clients in many ways, this SDK supports two of them:

* postMessage: a dApp will run in a container (`iframe` or mobile Apps' `webview`), and the container needs to overwrite `window.postMessage` method in the dApp, so the dApp and the container can communicate with each other by overwritten `postMessage`. AElf mobile wallet App has implemented this.
* WebSocket(Socket.io): use traditional B/S architecture, communicate by `WebSocket`. SDK uses `Socket.io` to support `WebSocket` communication, and this requires servers need to support `Socket.io` too. `aelf-command` has implemented this way.

Developers can choose one of them depending on requirements, in the process of development, we provide two ways to support data mock and debugging:

* [aelf-bridge-demo](https://github.com/AElfProject/aelf-bridge-demo): this demo uses `iframe` to overwrite `dapp.html`'s `postMessage` to simulate communication with mobile Apps.
* [aelf-command dapp-server](https://github.com/AElfProject/aelf-command): `aelf-command` provides a simple `socket.io` server to support the communication method `socket.io` in `aelf-bridge`, developers can change the communication way to `SOCKET.IO`, and give the URI given by running `aelf-command dapp-server` as an option when initializing `aelf-bridge` instance. Therefore, developers can inspect the communications in the Network tab of browser.

### Initialization

```javascript
import AElfBridge from 'aelf-bridge';

// Initialize the bridge instance, you can pass options during initialization to specify the behavior, see below for explanation
const bridgeInstance = new AElfBridge();
// init with options
const bridgeInstance = new AElfBridge({
  timeout: 5000 // ms
});

// After initializing the instance you need connect
bridgeInstance.connect().then(isConnected => {
  // isConnected True if the connection was successful.
})
```

#### Options

The options can be passed as follows:

```javascript
const defaultOptions = {
  proxyType: String, // The default is `POST_MESSAGE`. Currently, we support the `POST_MESSAGE` and `SOCKET.IO` proxy types are provided. The `Websocket` mechanism will be provided in the future. Valid values ​​are available via `AElfBridge.getProxies()`.
  channelType: String, // The default is `SIGN`, it is the serialization of the request and response, that is, Dapp exchanges the public and private keys with the client and the private key is used to verify the signature information, thereby verifying whether the information has been tampered with. Another method of symmetric encryption is provided. The parameter value is `ENCRYPT`, and the shared public key is used for symmetric encryption. The valid value of the parameter is obtained by `AElfBridge.getChannels()`.
  timeout: Number, // Request timeout, defaults to 3000 milliseconds
  appId: String, // The default is empty. Dapp does not specify if there is no special requirement. If you need to specify it, you need to randomly generate a 32-bit hex-coded id each time. A credential used to communicate with the client, specifying the Dapp ID. If it is not specified, the library will process it internally. The first run will generate a random 32-bit hex-encoded uuid. After the connection is successful, it will be stored in `localStorage`, then the value will be taken from `localStorage`. If not, then Generate a random id.
  endpoint: String, // The default is empty. If the address of the node is empty, the client uses the internally saved primary link address by default, and can also specify to send a request to a specific node.
  // Optional options in `POST_MESSAGE` communication mode
  origin: String, // The default is `*`, the second parameter of the `postMessage` function, in most cases you do not need to specify
  checkoutTimeout: Number, // The default is `200`, in milliseconds, it checks the client's injected `postMessage`. In most cases, you don't need to specify this
  urlPrefix: String, // The default is `aelf://aelf.io?params=`, which is used to specify the protocol and prefix of the node. If the client does not have special requirements, it does not need to be changed.
  // Optional options in `socket.io` communication mode.
  socketUrl: String, // The address of the websocket connection, the default is `http://localhost:50845`
  socketPath: String, // Path to the connection address, the default is empty
  messageType: String // Pass the type of the socket.io message, the default is `bridge`
}
```

### Get wallet account information

`bridgeInstance.account()`

```javascript
bridgeInstance.account().then(res => {
  console.log(res);
})
res = {
  "accounts": [
    {
      "name": "test",
      "address": "XxajQQtYxnsgQp92oiSeENao9XkmqbEitDD8CJKfDctvAQmH6"
    }
   ],
  "chains": [
    {
      "url": "http://13.231.179.27:8000",
      "isMainChain": true,
      "chainId": "AELF"
    },
    {
      "url": "http://52.68.97.242:8000",
      "isMainChain": false,
      "chainId": "2112"
    },
    {
      "url": "http://52.196.227.200:8000",
      "isMainChain": false,
      "chainId": "2113"
    }
  ]
}
```

### Calling the chain API

API for interacting with the node. The API available methods can be viewed by `{chain address}/swagger/index.html`, to get the currently supported APIs you can call `AElfBridge.getChainApis()`.

Developer can use the methods of `bridgeInstance.chain` to call the api.

Example:

* Get block height

```javascript
bridgeInstance.chain.getBlockHeight().then(console.log).catch(console.log)
```

* Get chain status
```javascript
bridgeInstance.chain.getChainStatus().then(console.log).catch(console.log)
```

### Call contract method (read-only and send transaction)

* Get contract method list
* Send transaction
* Contract read-only method

Example:

* Call the `Transfer` method of the `Token` contract to initiate a transfer transaction

```javascript
const tokenAddress = 'mS8xMLs9SuWdNECkrfQPF8SuRXRuQzitpjzghi3en39C3SRvf'; // Get contract address by genesis contract method `GetContractAddressByName`
bridgeInstance.chain.contractAt(tokenAddress).then(async contract => {
    const tokenInfo = await contract.GetTokenInfo.call({symbol: 'ELF'});
    const transactionId = await contract.Transfer({
          amount: "10000000000",
          to: "fasatqawag",
          symbol: "ELF",
          memo: "transfer ELF"
    });
    console.log(tokenInfo);
    console.log(transactionId);
})
```

### disconnect

Used to disconnect from the client and clearing the public key information, etc.

`bridgeInstance.disconnect()`


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.0.10](https://github.com/AElfProject/aelf-bridge/compare/v0.0.9...v0.0.10) (2022-03-15)


### Bug Fixes

* 🐛 fix data null ([e5e0a3d](https://github.com/AElfProject/aelf-bridge/commit/e5e0a3df0e750dddb3ea61df4ce735e190f11ae6))

### [0.0.9](https://github.com/AElfProject/aelf-bridge/compare/v0.0.8...v0.0.9) (2020-06-11)


### Features

* 🎸 isomorphism aelf-bridge and aelf-sdk ([a9b8d2d](https://github.com/AElfProject/aelf-bridge/commit/a9b8d2d87e9cd84e07eeff97188d1f7d6eaff5b1))

### [0.0.8](https://github.com/AElfProject/aelf-bridge/compare/v0.0.7...v0.0.8) (2019-12-13)


### Bug Fixes

* 🐛 fix error handle when receiving many messages ([6d08da6](https://github.com/AElfProject/aelf-bridge/commit/6d08da6ebd897d5e838a3532db984c49114794e8))

### [0.0.7](https://github.com/AElfProject/aelf-bridge/compare/v0.0.6...v0.0.7) (2019-12-09)


### Features

* 🎸 increase timeout for postMessage ([01923dc](https://github.com/AElfProject/aelf-bridge/commit/01923dc590fb63650db25d5233650ffd676e1705))

### [0.0.6](https://github.com/AElfProject/aelf-bridge/compare/v0.0.5...v0.0.6) (2019-12-03)


### Features

* 🎸 use HKDF to expand shared key ([8a349a2](https://github.com/AElfProject/aelf-bridge/commit/8a349a2a95be5736ae176ff7878b7ab49c3786be))

### [0.0.5](https://github.com/AElfProject/aelf-bridge/compare/v0.0.4...v0.0.5) (2019-11-21)


### Features

* 🎸 add timestamp for checking if request is expired ([02bafc8](https://github.com/AElfProject/aelf-bridge/commit/02bafc8468e054b233d6468487fa14e283afa702))

### [0.0.4](https://github.com/AElfProject/aelf-bridge/compare/v0.0.3...v0.0.4) (2019-11-08)

### [0.0.3](https://github.com/AElfProject/aelf-bridge/compare/v0.0.2...v0.0.3) (2019-11-04)


### Features

* 🎸 add proxy close function ([3c85891](https://github.com/AElfProject/aelf-bridge/commit/3c85891a90140bc9e105a1ab848f6eb30b617621))


### Bug Fixes

* 🐛 change when to close ([c896636](https://github.com/AElfProject/aelf-bridge/commit/c896636a3d218340b667b1e8d3602fb271b5291d))
* 🐛 rename params name ([5888237](https://github.com/AElfProject/aelf-bridge/commit/588823718f6f0f3480c714e13010f3d96cef1b8a))

### [0.0.2](https://github.com/AElfProject/aelf-bridge/compare/v0.0.1...v0.0.2) (2019-10-28)


### Features

* 🎸 add socket.io support ([159258b](https://github.com/AElfProject/aelf-bridge/commit/159258b559b72bfa4c31db5fd97099a8fc24e168))
* 🎸 exposed sendMessage method ([04c5c43](https://github.com/AElfProject/aelf-bridge/commit/04c5c43269fb2857161a1dd0e699143c470ad719))

### 0.0.1 (2019-10-22)


### Features

* 🎸 add initial commit ([ee757c6](https://github.com/AElfProject/aelf-bridge/commit/ee757c6ad9caf182301909fa80294b22f0663235))


src/index.js
/**
 * @file aelf isomorphism
 * @author atom-yang
 */
import MessageChannel from './messageChannel';
import Proxy from './proxy';
import {
  getUUIDForUrl
} from './utils/utils';
import StorageService from './utils/storage';
import {
  PROXY_TYPE,
  CHANNEL_TYPE,
  CHAIN_APIS,
  CHAIN_METHODS
} from './common/constants';
import Chain from './chain';
import ContractFactory from './contract';

const defaultOptions = {
  // common options
  proxyType: PROXY_TYPE.postMessage,
  channelType: CHANNEL_TYPE.sign,
  timeout: 30 * 60 * 1000,
  endpoint: '',
  // for post message
  origin: '*',
  checkoutInterval: 200,
  urlPrefix: 'aelf://aelf.io?params=',
  // for socket.io
  socketUrl: 'http://localhost:50845',
  socketPath: '',
  messageType: 'bridge'
};

export default class Bridge {
  static getProxies() {
    Proxy.getProxies();
  }

  static getChannels() {
    return MessageChannel.getChannels();
  }

  static getChainApis() {
    return CHAIN_APIS;
  }

  static getChainMethods() {
    return [...CHAIN_METHODS.map(v => v.methodName), 'contractAt'];
  }

  constructor(options = defaultOptions) {
    this.options = {
      ...defaultOptions,
      appId: StorageService.getAppId() || getUUIDForUrl(),
      ...options
    };
    this.request = new MessageChannel(this.options);
    this.connected = false;
    this.chain = new Chain(this);
    this.chain.contractAt = this.contractAt.bind(this);
  }

  connect() {
    return this.request.connect().then(res => {
      if (res) {
        StorageService.setAppId(this.options.appId);
      }
      return res;
    });
  }

  disconnect() {
    return this.request.disconnect().then(res => {
      this.connected = false;
      return res;
    });
  }

  async contractAt(address) {
    const list = await this.sendMessage('getContractMethods', {
      endpoint: this.options.endpoint,
      address
    });
    const r = new ContractFactory(address, list, this);
    return r.at();
  }

  async sendMessage(action, params) {
    if (!this.connected) {
      const connected = await this.connect();
      if (!connected) {
        throw new Error('Connection failed');
      }
      this.connected = true;
    }
    const res = await this.request.sendMessage({
      action,
      params
    });
    const {
      code,
      data
    } = res;
    if (+code === 0) {
      return data || 'null';
    }
    throw res;
  }

  invoke(params) {
    return this.sendMessage('invoke', {
      endpoint: this.options.endpoint,
      ...params
    });
  }

  invokeRead(params) {
    return this.sendMessage('invokeRead', {
      endpoint: this.options.endpoint,
      ...params
    });
  }

  api(params) {
    return this.sendMessage('api', {
      endpoint: this.options.endpoint,
      ...params
    });
  }

  account() {
    return this.sendMessage('account', {});
  }
}
